src/pushsource/_impl/helpers.py

```diff
@@ -107,3 +112,42 @@ def force_https(url):
     parsed = urlparse(url)
     with_https = ParseResult("https", *parsed[1:])
     return with_https.geturl()
+
+
+def as_completed_with_timeout_reset(futures, timeout):
+    """
+    Yields any finished future within time alloted by timeout, raises TimeoutError otherwise.
+    Futures are allowed to continue over timeout, if any processing in caller is slow.
+    Timeout is not applied to futures themselves but rather to wait() call, which means
+    that this function allows to yield all quicker futures and tries to wait for slow ones.
+
+    Parameters:
+        futures: List[Future]
+            list of futures
+
+        timeout: int or float
+            Timeout used for wait() calls
+
+    Returns:
+        Future
+            finished Future object
+
+    Throws:
+        TimeoutError, if waiting for some Future to finish in wait() call takes more
+        time than given timeout.
+    """
+    total_futures = len(futures)
+    pending = futures
+    while pending:
+        finished, pending = wait(pending, timeout=timeout, return_when=FIRST_COMPLETED)
+        # no future finished in timeout given, raise TimeoutError
+        if not finished:
+            # try to cancel futures, we can't cancel those already running or finished
```

Comment:
I think this cancel part doesn't fit, could we please drop it?  The rest looks good.

The reasons cancel doesn't seem right here are:
- the standard as_completed doesn't do this (and naming of this customized version doesn't give any hint there's a difference)
- it means the caller can't possibly decide to retry and wait for a bit longer on the current futures, because we're effectively "breaking" those futures before we return.
- it shouldn't be needed, because we should already be setting up our executors to cancel on shutdown.  So, if this raised TimeoutError here ends up being fatal, then we're already going to cancel all pending futures before allowing the exception to propagate to the caller of pushsource.
- unless I missed something, it wasn't covered by the tests.

Reply:
OK, I'll drop it.